types/parse5/index.d.ts

```diff
@@ -264,6 +262,11 @@ export interface DefaultTreeTextNode extends DefaultTreeNode {
  * Cast to the actual AST interface (e.g. {@link parse5.DefaultTreeNode}) to get access to the properties.
  */
 export type Node = DefaultTreeNode | object;
+/**
+ * Generic ChildNode interface.
+ * Cast to the actual AST interface (e.g. {@link parse5.DefaultTreeChildNode}) to get access to the properties.
+ */
+export type ChildNode = DefaultTreeChildNode | object;
```

Comment:
This all looks reasonable given the types already defined here, but nothing is _actually_ stricter because of this, right? Because `DefaultTreeChildNode | object` reduces to just `object` (unless the union is being used to check a fresh object literal, in which case it operates a little differently).

Reply:
Oh @andrewbranch i'm sorry, I had a lot of work to do on my main job, so completely forgot about this PR and now received the notification about it's merge.

The main idea is that:
1. now you can't pass `DefaultTreeDocumentType`  to `getParentNode`
2. now you have a generic type for all nodes that could have parent.

One more thing I have to fix here is type `childNodes`  property in `DefaultTreeParentNode` as `DefaultTreeChildNode` so this give an ability to write node traversal code like `node.childNodes[0].parentNode` (which return same node).

Reply:
> now you can't pass DefaultTreeDocumentType to getParentNode

But that’s not true. You can pass literally any object type, because that’s what `object` means.

<img width="650" alt="image" src="https://user-images.githubusercontent.com/3277153/61552498-d6b49880-aa0c-11e9-8756-45571087582a.png">

All these `| object` basically make all the other typings meaningless. I merged this because that was already the case for _everything_ in these typings, but if you want to fix them in the future, you need to get rid of those union types.

Reply:
And by the way, if you intend to ensure that, e.g., “now you can't pass DefaultTreeDocumentType to getParentNode,” you should add that as a test case with `// $ExpectError`, so you (and reviewers here) don’t have to guess at what the effect of a PR is.

Reply:
@andrewbranch thank you so much for this comment, I'm not that familiar with `ts` and I was banging my head on this for a long time.

Do you see any reason why the `| object` is even there? Can it be removed? Not only it creates the case that you described in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36817/files#r305446160, but using these typedefs is completely meaningless as typescript always thinks that anything returned by parser is just `object` with no interface props available 😭 

<img width="567" alt="image" src="https://user-images.githubusercontent.com/5036933/62240990-80663300-b3d8-11e9-82e2-69375afecf01.png">

Reply:
Sorry, I'm not familiar enough with parse5 to know if this would break anyone or not :/

It definitely seems wrong, but it's possible that it's wrong for a _reason_. This is kind of a weird package because the JS is used by millions of people every day (it’s a dependency of jsdom, notably), but relatively few people are depending on the types. Still, the types package has some popular-ish dependencies, so my advice would be to test whether your changes break some of the more popular ones there: https://www.npmjs.com/browse/depended/@types/parse5

A quick sampling shows these are fairly popular:

- dom5
- polymer-analyzer
- polyserve

Reply:
Thanks for some guidance! I'll try looking into it!